includes/plugins/wc-memberships/class-memberships.php

```diff
@@ -1086,5 +1087,101 @@ public static function handle_wc_memberships_expire_user_membership( $expire, $u
 		}
 		return $expire;
 	}
+
+	/**
+	 * When a subscription's status changes, ensure that any related user memberships are linked to the correct subscription.
+	 *
+	 * @param WC_Subscription $subscription An instance of a WC_Subscription object that just had its status changed.
+	 */
+	public static function check_user_memberships_on_subscription_update( $subscription ) {
+		self::maybe_relink_user_membership_subscription( $subscription->get_user_id() );
+	}
+
+	/**
+	 * When a reader logs in, ensure that their user memberships are linked to the correct subscription.
+	 *
+	 * @param string  $user_login User's login.
+	 * @param WP_User $user WP_User object for the logged-in user.
+	 */
+	public static function check_user_memberships_on_login( $user_login, $user ) {
+		if ( ! Reader_Activation::is_user_reader( $user ) ) {
+			return;
+		}
+		self::maybe_relink_user_membership_subscription( $user->ID );
+	}
+
+	/**
+	 * Ensure that the user membership is linked to the user's most recent active subscription, if any.
+	 *
+	 * @param int $user_id User ID.
+	 *
+	 * @return bool True if user membership was relinked to a different subscription, false otherwise.
+	 */
+	public static function maybe_relink_user_membership_subscription( $user_id ) {
+		$updated = false;
+		if ( ! self::is_active() ) {
+			return $updated;
+		}
+		$user_memberships = wc_memberships_get_user_memberships( $user_id );
+		foreach ( $user_memberships as $user_membership ) {
+			$membership_plan_id      = $user_membership->get_plan_id();
+			$user_membership_id      = $user_membership->get_id();
+			$subscription_membership = new \WC_Memberships_Integration_Subscriptions_User_Membership( $user_membership_id );
+			$active_subscription_id     = self::get_user_subscription_for_membership_plan( $user_id, $subscription_membership->get_plan_id() );
+			if ( $subscription_membership && ! empty( $active_subscription_id ) ) {
+				$linked_subscription_id = (int) $subscription_membership->get_subscription_id();
+
+				// If the user membership is linked to the wrong subscription, relink it.
+				if ( $linked_subscription_id !== $active_subscription_id ) {
+					$updated         = $subscription_membership->set_subscription_id( $active_subscription_id );
+					$membership_plan = $subscription_membership->get_plan();
+
+					// Reset end date for plans with access set to subscription length.
+					if ( $membership_plan->is_access_length_type( 'subscription' ) && ! empty( $subscription_membership->get_end_date() ) ) {
+						$subscription_membership->set_end_date( '' );
+					}
```

Comment:
With the hook priority at the default (10), the `set_subscription_id()` works and the membership is relinked, but `set_end_date()` doesn't persist and the membership remains with the expired status.

After changing the hook priority to 11 the end date resets, but the status remains "expired". Maybe we should be running a `$subscription_membership->update_status( 'active', $note );` here too?

Reply:
f436f1a is now working for me. As for whether we still need `handle_wc_memberships_expire_user_membership()`, I repurposed it as another trigger to check and possibly relink the membership, which should prevent the "Expired" status from being set if the membership is linked to the wrong subscription. WDYT?

Reply:
3fad077 updates `handle_wc_memberships_expire_user_membership()` to `check_user_memberships_on_expire()` and groups it with the other `check_user_memberships_on_*` methods, for clarity. It also only prevents the expiration if that specific membership should be relinked to an active subscription.